### ISSUE TITLE
Remote rapp indexer

### ISSUE THREAD

**Issue by bit-pirate** [robotics-in-concert/rocon_app_platform#189]
Develop a module, which queries remote repositories for available rapps and stores their information (.rapp, .interface, .launch, image files) in an index.

**Comment by bit-pirate** [comment]
Some design questions and comments:
- Handling image files
  - download them into a temporary folder on index generation?
  - store links (URL) to the images and use a placeholder, when URL is invalid/not reachable?
- Rapp versioning
  - If rapps change, put them into different branch in the rapp repo pointed to
- Caching
  - would like to have 2 options:
    - do the indexing using a cache (reduces indexing time for users)
  - allow (triggering of) complete re-indexing (reduces waiting time for developers)
- 1 rapp/package?
  - Not needed. Multiple rapps per package can be supported.

**Comment by esteve** [comment]
One of the differences between the remote and local indexer is that we can't easily crawl remote repositories (not all web repositories support directory listings).

We can either:
- pack rapps and their resources into single-file archives
- generate a file listing the file names of the rapp resources

both approaches add an extra step from the user before publishing a rapp. I favor the first, because it simplifies fetching rapps and separates clearly a published rapp from the rapp itself. 

I see rapps akin to Gentoo repositories or BSD ports. That is, rapps are metadata on top of ROS packages, so we could distribute entire rapps repositories, and only when we want to install the rapps locally, the dependencies are resolved and the ROS packages fetched and installed. This would simplify the design heavily, and we could reuse the local indexer, which is already implemented.

**Comment by bit-pirate** [comment]
> We can either:
> - pack rapps and their resources into single-file archives
> - generate a file listing the file names of the rapp resources

I believe the first option is not feasible. The *.rapp and *.interface file could easily be merged. Adding the launcher however would make things messy and more difficult. However, we won't be able the merge the image (rapp icon). So in the end we would have multiple files as well.

I'm fear I don't fully understand the existing problems when it comes to directory listing. Could you explain a bit further what exactly is difficult to do? The rapp currently contains up to 4 files in one single directory. If a pointer to this directory is provided, is it still difficult to access all files in this folder?

> I see rapps akin to Gentoo repositories or BSD ports. That is, rapps are metadata on top of ROS packages, so we could distribute entire rapps repositories, and only when we want to install the rapps locally, the dependencies are resolved and the ROS packages fetched and installed. This would simplify the design heavily, and we could reuse the local indexer, which is already implemented.

That's how we see rapps as well.

**Comment by esteve** [comment]
> I'm fear I don't fully understand the existing problems when it comes to directory listing. Could you explain a bit further what exactly is difficult to do? The rapp currently contains up to 4 files in one single directory. If a pointer to this directory is provided, is it still difficult to access all files in this folder?

If a developer stores their Rapps on Github by just uploading the Rapp's files, for example, we can't depend on server-provided directory listings:

https://raw.githubusercontent.com/robotics-in-concert/rocon_app_platform/hydro-devel/rocon_apps/apps/chirp/

so, we either add an extra step before releasing a rapp to generate a list of a rapp's contents so they can be fetched individually or we archive the rapp and upload it instead.

My proposal is that we just create a tarball/zip of a rapp's (or a list of rapps) contents and let the Rapp client fetch the archives. And, if we allow multiple rapps to be archived into a single file, in the end, pointing the client to an archived rapp or to an index become one and the same. The ports collection in BSD store all the metadata locally (including for packages that are not currently installed), and it's only when a port gets installed that its dependencies are fetched.

**Comment by bit-pirate** [comment]
> My proposal is that we just create a tarball/zip of a rapp's (or a list of rapps) content

Hm, tarballs are not that nice, since they are binary blobs and are not placed well inside a source code repo.

What do you think about this approach:

Each rapp repo contains its own rapp index. That is, a text file, which contains all information [1] about the rapps in that repo. The remote rapp indexer - equipped with the URLs of those index files aka index chunks - can then directly fetch those files and piece them together to create one big index of local and remote rapps. Those rapp repo indexes could be created by a script, which the rapp developer runs locally on his machine after he finished his rapp development. He then commits his rapp changes and the index to his repo. Adding an index to an rapp repo would then be kind of a "release".

---

[1] images would still be a problem; but identifying the URL to the image should be feasible, e.g. rapp repo index URL + relative URL to the image:

https://raw.githubusercontent.com/robotics-in-concert/rocon_app_platform/hydro-devel/rocon_apps/in

**Comment by esteve** [comment]
Sounds good, I'm going to write a simple fetcher and an index generator and we can iterate over it.

**Comment by bit-pirate** [comment]
Quoting @esteve 's follow-up question from PR #196:

> Things to discuss:
> 
> Where should Rapps be installed?

Not necessary, since all the relevant information is stored in the index. In case you are asking, where the rapp index should be stored, then I would suggest to put in  `~/.ros/rocon/`.

> How to handle versioning?

My suggestion posted at the beginning: Handle different versions by putting them in different branches.

> Download Rapps individually or by package?

By rapp repo I would say, since this is where the indexer is pointed to (assuming you are talking about the index, instead of the rapp)

> Any other information that should be in the index file?

So far only what the rapps contain: .rapp, .interface, .launch, _image_ files

> Should an index store information about all the Rapps in a tree instead of separate indices?

<del>I think separate indices would be fine. In case of the tree, how would its structure look like?</del>

After seeing the result of your current implementation, I now understand this question better. As mentioned above, I think the generation of one big index would be simpler. If this index is a tree with just one level or multiple can still be discussed - I think the first would be sufficient.

For developers, who want to create and index for their repo, it would be good to offer them the option to point to a folder/repo/package, which is then crawled for rapps. Then index would then also be placed in that path, e.g.:

`$ rocon_app index ~/rocon_ws/src/rocon_app_platform` will gather all information on discovered rapps in `~/rocon_ws/src/rocon_app_platformrapp.index`.

Where as `rocon_app index` will create a global index at `~/.ros/rocon/rapp.index` containing all locally and remotely found rapps.

What do you think about that?

**Comment by esteve** [comment]
> Quoting @esteve 's follow-up question from PR #196:
> 
> Things to discuss:
> 
> Where should Rapps be installed?
> 
> Not necessary, since all the relevant information is stored in the index. In case you are asking, where the rapp index should be stored, then I would suggest to put in ~/.ros/rocon/.

+1

> How to handle versioning?
> 
> My suggestion posted at the beginning: Handle different versions by putting them in different branches.

+1

> Download Rapps individually or by package?
> 
> By rapp repo I would say, since this is where the indexer is pointed to (assuming you are talking about the index, instead of the rapp)

+1

> Any other information that should be in the index file?
> 
> So far only what the rapps contain: .rapp, .interface, .launch, image files

Does the index actually contain the content for those files or just
point to a path to them?

> After seeing the result of your current implementation, I now understand this question better. As mentioned above, I think the generation of one big index would be simpler. If this index is a tree with just one level or multiple can still be discussed - I think the first would be sufficient.

+1

If we have one big index, I think it goes back to my earlier proposal
where all the Rapp information is stored in one single place and Rapp
client just fetches this one big index.

> For developers, who want to create and index for their repo, it would be good to offer them the option to point to a folder/repo/package, which is then crawled for rapps. Then index would then also be placed in that path, e.g.:
> 
> $ rocon_app index ~/rocon_ws/src/rocon_app_platform will gather all information on discovered rapps in ~/rocon_ws/src/rocon_app_platformrapp.index.
> 
> Where as rocon_app index will create a global index at ~/.ros/rocon/rapp.index containing all locally and remotely found rapps.
> 
> What do you think about that?

Sounds good. I actually added an argument to rocon_app index to
generate an index from a given tree, but I left it unimplemented just
so we had some code we could discuss:

https://github.com/robotics-in-concert/rocon_app_platform/pull/196/files#diff-c11fcec18d33c6d7f92bb796c0b0b57fR161

**Comment by bit-pirate** [comment]
> Does the index actually contain the content for those files or just
> point to a path to them?

I would suggest to actually put those contents into the index. Otherwise all lot of downloads may be required during runtime, what will slow down the starting of a rapp.

Ping me, when there is more to test.

**Comment by esteve** [comment]
So, at that point, I think we could just store the Rapps contents into tarballs and release them. The client would fetch them, unpack them and merge the local index with the index in the downloaded tarball.

**Comment by bit-pirate** [comment]
My do you want to use tarballs?

Please review my comment on your previous suggestion of using tarballs further up (https://github.com/robotics-in-concert/rocon_app_platform/issues/189#issuecomment-39838518) and let me know, how tarballs would do the job better (and amything else I might have missed).

**Comment by esteve** [comment]
I'm probably missing some details, so sorry if I seem obtuse, but it looks to me that we want to have all the Rapp resources together (images, .rapp, .launch and .interface files). Using tarballs (or any other packing format) would just put the resources in a single place, and that way we can not only separate a Rapp development files from released Rapps, but also simplify distribution and handle repository versioning (e.g. we can have snapshots of a repository).

When I mentioned BSD ports, I saw that Rapps as metadata for ROS packages are analogous to ports Makefiles for system packages. For example, the FreeBSD ports collection (https://www.freebsd.org/ports/installing.html) contains all the metadata to build and install a huge array of system packages, even though the collection itself is distributed as a tarball (ftp://ftp.freebsd.org/pub/FreeBSD/ports/ports/ports.tar.gz) which weighs less than 50Mb. This doesn't mean that all the system packages are in there, just the metadata to build and install them. I think the Rapps case is very similar, Rapps provide the metadata to install ROS packages, we can distribute a Rapps repository as a tarball so users have the metadata locally to install ROS packages, which is provided in the form of Rapps.

**Comment by stonier** [comment]
Jumping into the middle of this so forgive me if I go astray on a couple of things. Taking a step back here - what we were **initially envisioning**:
- App manager is fed
  - ROS_PACKAGE_PATH which it locally crawls git cloned rapp repos (easy to use while developing)
  - A param pointing to an indexed git rapp repo branch which has an index file (with a known name) stored at its root.
- Versioning/releasing rapp repos simply handled by pointing at different tags or branches.

I've taken a step back though and started digging around **ports** and the idea of a **tarball** to see what it looks like. Is this in line with what you were thinking @esteve?
- Dump all metadata and content (e.g. icons) for all rapps in the repo into the tarball.
  - Esteve's opinion is that it shouldn't really get too large to be unwieldy, even with icons (c.f. bsd ports).
- App manager on startup checks to see if it has an up to date tarball already downloaded for that repo
  - If not, it downloads
- Versioning by tarball url.

If this is right, I think I'm actually getting swayed by this (yeah, I'm such a shallow and impressionable fellow!). My concern that downloading too many icons we  never run might be negligible (at least for the near future) and if we transfer all rapp meta/icon downloads to app manager startup (not rapp start), then it's not going to slow down the startup process. Additionally, it will probably download the first time you run the rapp manager, but shouldn't have to redownload that often afterwards. You could even use this with offline robots - just dump the tarballs in the home dir. What do think @bit-pirate?

There is also a separate question of what to do with locally and remotely known indexes. I think:
- Locally generated indexes at startup need not get saved/cached anywhere.
  - There is the problem of working out which runtime ROS_PACKAGE_PATH corresponds to which cache
  - It's not too much trouble getting an app manager to regenerate this on startup every time.
- Remote index can be saved in `~/.ros/rocon` and checked for new updates (like gradle/maven) on startup as indirectly inferred above.

**Comment by stonier** [comment]
Icons probably need a bit of a rework. Currently they use ros resource names (e.g. `rocon_icons/foo.png`) to be discovered. i.e. we find them in some package on the `ROS_PACKAGE_PATH`. 

If we download them in an index _with_ the metadata, then that probably won't work. Some options:
- on the ROS_PACKAGE_PATH as a ros resource name to a _dependent_ package that will of course get installed when the rapp is installed.
- url to a wget'table resource
- relative path location (from the .rapp file)
- forced side by side with the rapp file.

_Ros resource names_ are a problem - they'll not be found if someone (e.g. IDE, not app manager) just wants to visually view the index. Url's are weak (consider _offline_ app managers/IDE's) unless the indexer is tasked with the job of going and retrieving the url when it makes the index. Thoughts?

**Comment by bit-pirate** [comment]
So, I sat down with @stonier to review and discuss the benefits and drawbacks of the text-file and tarball approaches.

Compared to text files, tarballs require a certain extra time for access/extraction. I personally don't know exact numbers and @stonier sees this extra time as neglectable. If you agree, then I'm happy to drop this concern. Another drawback is, that those tarballs will often be stored in a VCS, which is not the best place to put them. However, since the tarballs won't be large and changed often, we think this can be neglected as well.

The big benefit of tarballs would be that they can contain the image files (icons) as well, what removes the need to download all rapp icons individually. 

In summary we support the usage of tarballs. So, please go ahead! :ok_hand: 

Last, but not least a follow-up on the icon location concern:
We like to drop the usage of ROS resource names (`<package name>`/`<filename>`) due to the issues @stonier mentioned in the previous post. Instead we like to enforce the placement of icon files alongside the rapp file.

**Comment by esteve** [comment]
> I've taken a step back though and started digging around ports and the idea of a tarball to see what it looks like. Is this in line with what you were thinking @esteve?

Yep, that's it. The BSD folks have a lot of experience with this sort of packaging, and I think their solution is very elegant and simple (the current FreeBSD ports collections includes recipes for over 23000 packages).

The release process becomes rather simple, we just package all the Rapps in a directory into a tarball and serve it. As for incremental updates, we can either let the client redownload the Rapps collection, or we could use zsync (http://zsync.moria.org.uk/), but I don't think the Rapps tarball will  weigh too much, the FreeBSD ports collections, with over 23000 packages just weighs 49M after all.
